### PR TITLE
gobject-introspection: rebuild against 3.13

### LINKED ISSUE #37369: at-spi2-core fails to build on amd64
The package at-spi2-core fails to build from source with the following error:

```
[204/292] Compiling C object atk-adaptor/libatk-bridge-2.0.so.0.0.0.p/meson-generated_.._.._xml_introspection.c.o
[205/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/value-adaptor.c.o
[206/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/text-adaptor.c.o
[207/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/table-cell-adaptor.c.o
[208/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/table-adaptor.c.o
[209/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/socket-adaptor.c.o
[210/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/selection-adaptor.c.o
[211/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/image-adaptor.c.o
[212/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/hypertext-adaptor.c.o
[213/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/hyperlink-adaptor.c.o
[214/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/editabletext-adaptor.c.o
[215/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/document-adaptor.c.o
[216/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/component-adaptor.c.o
[217/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/collection-adaptor.c.o
[218/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/cache-adaptor.c.o
[219/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/application-adaptor.c.o
[220/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/action-adaptor.c.o
[221/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/accessible-adaptor.c.o
[222/292] Compiling C object atk-adaptor/adaptors/libatk-bridge-adaptors.a.p/meson-generated_.._.._.._xml_introspection.c.o
[223/292] Compiling C object droute/droute-test.p/droute-test.c.o
[224/292] Compiling C object registryd/at-spi2-registryd.p/ucs2keysym.c.o
[225/292] Compiling C object registryd/at-spi2-registryd.p/event-source.c.o
[226/292] Compiling C object registryd/at-spi2-registryd.p/display.c.o
[227/292] Compiling C object registryd/at-spi2-registryd.p/deviceeventcontroller-x11.c.o
[228/292] Compiling C object registryd/at-spi2-registryd.p/registry.c.o
[229/292] Compiling C object registryd/at-spi2-registryd.p/registry-main.c.o
[230/292] Compiling C object registryd/at-spi2-registryd.p/reentrant-list.c.o
[231/292] Compiling C object registryd/at-spi2-registryd.p/marshal-dbus.c.o
[232/292] Linking static target tests/at-spi2-atk/libxmlloader.a
[233/292] Linking static target tests/at-spi2-atk/dummyatk/libdummyatk.a
[234/292] Compiling C object registryd/at-spi2-registryd.p/deviceeventcontroller.c.o
[235/292] Compiling C object registryd/at-spi2-registryd.p/meson-generated_.._.._xml_introspection.c.o
[236/292] Compiling C object atspi/libatspi.so.0.0.1.p/.._dbind_dbind-any.c.o
[237/292] Compiling C object atspi/libatspi.so.0.0.1.p/.._dbind_dbind.c.o
[238/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-device-x11.c.o
[239/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-value.c.o
[240/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-text.c.o
[241/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-table-cell.c.o
[242/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-table.c.o
[243/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-stateset.c.o
[244/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-selection.c.o
[245/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-relation.c.o
[246/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-registry.c.o
[247/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-object.c.o
[248/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-misc.c.o
[249/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-matchrule.c.o
[250/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-image.c.o
[251/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-hypertext.c.o
[252/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-hyperlink.c.o
[253/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-gmain.c.o
[254/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-event-listener.c.o
[255/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-editabletext.c.o
[256/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-document.c.o
[257/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-device-listener.c.o
[258/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-device-legacy.c.o
[259/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-device.c.o
[260/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-component.c.o
[261/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-collection.c.o
[262/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-application.c.o
[263/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-action.c.o
[264/292] Linking static target tests/at-spi2-atk/libtestutils.a
[265/292] Compiling C object atspi/libatspi.so.0.0.1.p/atspi-accessible.c.o
[266/292] Compiling C object atspi/libatspi.so.0.0.1.p/meson-generated_.._atspimarshal.c.o
[267/292] Generating GObject enum file atspi/atspi-enum-types.c (wrapped by meson to capture output)
[268/292] Compiling C object atk/libatk-1.0.so.0.25409.1.p/meson-generated_.._atk-enum-types.c.o
[269/292] Linking static target atk-adaptor/adaptors/libatk-bridge-adaptors.a
[270/292] Linking target atk/libatk-1.0.so.0.25409.1
[271/292] Compiling C object atspi/libatspi.so.0.0.1.p/meson-generated_.._atspi-enum-types.c.o
[272/292] Linking target atspi/libatspi.so.0.0.1
cc: warning: linker hardening options not enabled by '-fhardened' because other link options were specified on the command line
cc: warning: linker hardening options not enabled by '-fhardened' because other link options were specified on the command line
cc: warning: linker hardening options not enabled by '-fhardened' because other link options were specified on the command line
[273/292] Generating atk/Atk-1.0.gir with a custom command (wrapped by meson to set env)
[274/292] Generating symbol file atk/libatk-1.0.so.0.25409.1.p/libatk-1.0.so.0.25409.1.symbols
[275/292] Linking target tests/atk/testvalue
[276/292] Linking target tests/atk/teststateset
[277/292] Linking target tests/atk/testrelation
[278/292] Linking target tests/atk/testrole
[279/292] Linking target tests/atk/testdocument
Traceback (most recent call last):
  File "/usr/bin/g-ir-scanner", line 103, in <module>
    from giscanner.scannermain import scanner_main
  File "/usr/lib/gobject-introspection/giscanner/scannermain.py", line 35, in <module>
    from giscanner.ast import Include, Namespace
  File "/usr/lib/gobject-introspection/giscanner/ast.py", line 27, in <module>
    from .sourcescanner import CTYPE_TYPEDEF, CSYMBOL_TYPE_TYPEDEF
ninja: job failed: env PKG_CONFIG_PATH=/home/build/output/meson-uninstalled PKG_CONFIG=/usr/bin/pkg-config CC=cc /usr/bin/g-ir-scanner --quiet --no-libtool --namespace=Atk --nsversion=1.0 --warn-all --output atk/Atk-1.0.gir --quiet --c-include=atk/atk.h -DATK_COMPILATION -I/home/build/atk -I/home/build/output/atk -I/home/build/. -I/home/build/output/. -I/home/build/atk -I/home/build/output/atk --filelist=/home/build/output/atk/libatk-1.0.so.0.25409.1.p/Atk_1.0_gir_filelist --include=GObject-2.0 --symbol-prefix=atk --identifier-prefix=Atk --pkg-export=atk --cflags-begin -D_POSIX_C_SOURCE=200809L -D_DEFAULT_SOURCE -DG_DISABLE_SINGLE_INCLUDES -DATK_DISABLE_SINGLE_INCLUDES -I/home/build/. -I/home/build/output/. -I/home/build/atk -I/home/build/output/atk -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/gobject-introspection-1.0 --cflags-end --add-include-path=/usr/share/gir-1.0 -L/home/build/output/atk --library atk-1.0 -lglib-2.0 -lgobject-2.0 -lgirepository-1.0
  File "/usr/lib/gobject-introspection/giscanner/sourcescanner.py", line 31, in <module>
    from giscanner._giscanner import SourceScanner as CSourceScanner
ModuleNotFoundError: No module named 'giscanner._giscanner'
cc: warning: linker hardening options not enabled by '-fhardened' because other link options were specified on the command line
cc: warning: linker hardening options not enabled by '-fhardened' because other link options were specified on the command line
cc: warning: linker hardening options not enabled by '-fhardened' because other link options were specified on the command line
ninja: subcommand failed
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /usr/bin/ninja -C /home/build/output -j 64
deleting guest dir /tmp/melange-guest-2540262391
deleting workspace dir /tmp/melange-workspace-4023041230
removing image path /tmp/melange-guest-3789010655
failed to build package: unable to run package at-spi2-core pipeline: unable to run pipeline: unable to run pipeline: exit status 1
```

### PR DESCRIPTION
**Pull request by sil2100**
We see a FTBFS of `at-spi2-core`. The main reason is that `gobject-introspection` was built against 3.12, and the build can't work with 3.13. Rebuild it and also switch to using the py-versioned versions of packages.

Fixes: https://github.com/wolfi-dev/os/issues/37369

Related: https://github.com/wolfi-dev/os/issues/37197

### CODE REVIEW

File: gobject-introspection.yaml [hunk 1]
```diff
@@ -40,8 +43,9 @@ environment:
       - meson
       - openssf-compiler-options
       - posix-libc-utils
-      - py3-packaging
-      - py3-setuptools
+      - py${{vars.py-version}}-build-base
+      - py${{vars.py-version}}-packaging
+      - py${{vars.py-version}}-setuptools
       - python3
       - python3-dev
```
Comment (smoser) [inline gobject-introspection.yaml:50]: drop `python3` and `python3-dev` here. they're either already covered or potentially bring the wrong thing in.

actually, i think just do `py${{vars.py-version}}-build-base-dev` and you'll get what you need.

File: gobject-introspection.yaml [hunk 2]
```diff
@@ -76,7 +80,7 @@ subpackages:
         - gobject-introspection
         - libtool
         - posix-libc-utils
-        - py3-setuptools
+        - py${{vars.py-version}}-setuptools
         - python3
```
Comment (smoser) [inline gobject-introspection.yaml:84]: drop `python3` here. melange will add the necessary depends for it. (on python-3.13-base).

**Review by smoser — CHANGES_REQUESTED** [review]
thanks. please feel free to ask for clarification, or to disagree.

**Comment by murraybd** [comment]
I have a much worse PR in https://github.com/wolfi-dev/os/pull/37370 but there is one thing I was adding that is missed here. There is a binary `g-ir-scanner` which is not able to run because a runtime dependency on python3 did not exist.

I did some experimenting using your PR and adding python3.13 as a run time dependency but that did not resolve the issue because the shbang in `g-ir-scanner` only ever has python.

```
(20/20) Installing gobject-introspection (1.82.0-r2)
OK: 77 MiB in 34 packages
2664dae60eac:/work/packages# head -n1 /usr/bin/g-ir-scanner 
#!/usr/bin/env python3
```

**Comment by dannf** [comment]
> I have a much worse PR in #37370 but there is one thing I was adding that is missed here. There is a binary `g-ir-scanner` which is not able to run because a runtime dependency on python3 did not exist.
> 
> I did some experimenting using your PR and adding python3.13 as a run time dependency but that did not resolve the issue because the shbang in `g-ir-scanner` only ever has python.
> 
> ```
> (20/20) Installing gobject-introspection (1.82.0-r2)
> OK: 77 MiB in 34 packages
> 2664dae60eac:/work/packages# head -n1 /usr/bin/g-ir-scanner 
> #!/usr/bin/env python3
> ```
> 
> The point being if you continue working on this please sort out getting `g-ir-scanner` to work. Thanks!

This package has a [build-time config for the python interpreter](https://gitlab.gnome.org/GNOME/gobject-introspection/-/blob/1.82.0/meson_options.txt?ref_type=tags#L21) - but even if you explicitly set it to `python3.13`, it will [still shave the `.13` off in the shebang](https://gitlab.gnome.org/GNOME/gobject-introspection/-/blob/1.82.0/tools/meson.build?ref_type=tags#L15). I don't *think* `python-as-wrapper`, would help here because it explicitly mangles the path, but I could be wrong.

I submitted this upstream to add a shebang option:
https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/510

For now, we may just need to do a `sed` of the shebangs in the destdir.

File: gobject-introspection.yaml (hunk 1 above)
Comment (sil2100) [reply gobject-introspection.yaml:50]: Ah, I need to one day sit down and just browse through all the packages available for the Python multi-versioned efforts. Forgot about `-build-base-dev`, convenient!

**Comment by smoser** [comment]
> For now, we may just need to do a sed of the shebangs in the destdir.

we should have a pipeline like below. There are a _lot_ of places where we do shbang rewrites and all very inconsistently.

```yaml
uses: shbang/edit
with:
  # matches match whole line (^<match>$), the ^ and $ added.
  matches: ".*python3"
  new: "/usr/bin/python${{vars.py-version}}"
```

**Comment by smoser** [comment]
https://github.com/wolfi-dev/os/pull/37500 . i'm going to close this. Please feel free to re-open if I should not have.

**Pull request closed without merging**